crates/wdl-engine/src/cache.rs

```diff
@@ -0,0 +1,635 @@
+                .await
+                .context("failed to join lock task")?
+            }
+            Err(TryLockError::Error(e)) => Err(e).with_context(|| {
+                format!(
+                    "failed to acquire shared lock on cache entry file `{path}`",
+                    path = path.display()
+                )
+            }),
+        }
+    }
+
+    /// Acquires an exclusive file lock for the given path.
+    ///
+    /// If the file does not exist, it is created.
+    ///
+    /// If the file exists, it is truncated once the lock is acquired.
+    async fn acquire_exclusive(path: &Path) -> Result<Self> {
+        // Create or open the file, but do not truncate it if it exists before the lock
+        // is acquired
+        let mut options = fs::OpenOptions::new();
+        options.create(true).write(true);
+        let file = options.open(path).with_context(|| {
+            format!(
+                "failed to create call cache entry file `{path}`",
+                path = path.display()
+            )
+        })?;
+
+        let file = match file.try_lock() {
+            Ok(_) => Ok(Self(file)),
+            Err(TryLockError::WouldBlock) => {
+                let path = path.to_path_buf();
+                spawn_blocking(move || {
+                    info!(
+                        "waiting to acquire exclusive lock on cache entry file `{path}`",
+                        path = path.display()
+                    );
+
+                    file.lock().with_context(|| {
+                        format!(
+                            "failed to acquire exclusive lock on cache entry file `{path}`",
+                            path = path.display()
+                        )
+                    })?;
+                    Ok(Self(file))
+                })
+                .await
+                .context("failed to join lock task")?
+            }
+            Err(TryLockError::Error(e)) => Err(e).with_context(|| {
+                format!(
+                    "failed to acquire exclusive lock on cache entry file `{path}`",
+                    path = path.display()
+                )
+            }),
+        }?;
+
+        file.set_len(0).with_context(|| {
+            format!(
+                "failed to truncate cache entry file `{path}`",
+                path = path.display()
+            )
+        })?;
+
+        Ok(file)
+    }
+}
+
+impl Deref for LockedFile {
+    type Target = File;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for LockedFile {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl Read for LockedFile {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        self.0.read(buf)
+    }
+}
+
+impl Write for LockedFile {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.0.write(buf)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.0.flush()
+    }
+}
```

Comment:
This would go nicely in its own module to reduce the size of this one, perhaps also making the error messages less cache-specific so the type could be used for other purposes.